File: Review-15-Python-Strategy-Pattern/Example-2/examples/game.py
"""
This module handles the top-level game logic, including...

  1. Player symbol assignment
  2. Setting up the Referee
  3. Tracking match progress
  4. Managing player move order and validation
"""

from enum import StrEnum, auto
from typing import Never, Optional

from .board import Board
from .player import Player
from .referee import Referee


class GameStateError(Exception):
    """
    This exception is raised if a game is started before both players have been
    added and their symbols set.
    """


class GameState(StrEnum):
    """
    Used to specify the current state of the game (e.g., not started, in
    progress, or complete)
    """

    NOT_STARTED = auto()
    IN_PROGRESS = auto()
    OVER_WITH_STALEMATE = auto()
    OVER_WITH_WIN = auto()


class Game:
    """
    Orchestrates a single match of Tic-Tac-Toe.
    """

    def __init__(self):
        self._board = Board()
        self._ref = Referee(self._board)

        self._player1 = None
        self._player2 = None

        self._winner = None
        self._loser = None

    def set_players(self, player1: Player, player2: Player) -> None:
        """
        Setting player 1 and player 2 and adjust their symbols.

        Args:
            player1: the 'X' player
            player2: the 'O' player
        """

        self._player1 = player1
        self._player2 = player2

        self._player1.set_symbol("X")
        self._player2.set_symbol("O")

    def play_match(self) -> None:
        if not self.ready_to_start():
            #  return
            raise GameStateError("Player 1 *and** player 2 must be added")

        while self.is_not_over():
            self.play_round()

        print(self.get_board())

        if self.ended_with_win():
            print(f"Congratulations {self.get_winner()}!")

    def play_round(self) -> bool:
        """
        Play one round of Tic-Tac-Toe.

        Returns:
            True if the game ended during the round

        Raises:
            IOException if there is an error reading the selected move
        """

        # The game ended already
        if self._board.is_full():
            return True

        winner_id = 0

        print()
        print(self._board)
        self._round_turn(self._player1)

        # The game is over
        if self._board.is_full():
            print(self._board)

            return True

        winner_id = self._ref.check_for_win()

        if winner_id == 1:
            self._winner = self._player1
            self._loser = self._player2

            return True

        print()
        print(self._board)
        self._round_turn(self._player2)

        # Final board
        print()
        print(self._board)

        winner_id = self._ref.check_for_win()

        if winner_id == 2:
            self._winner = self._player2
            self._loser = self._player1

            return True

        return False

    def _round_turn(self, player) -> bool:
        """
        Get a player move, and update the board.

        If a player provides an invalid move... keep reprompting until a valid
        move is provided.
        """
        move = player.next_move()

        while not self._ref.selected_cell_is_empty(move):
            move = player.next_move()

        sym = player.get_symbol()

        self._board.set_cell(move, sym)

        return True

    # ---------------------------------------------------------------------------
    # Game Component (board and player) accessors
    # ---------------------------------------------------------------------------

    def get_player1(self) -> Player:
        return self._player1

    def get_player2(self) -> Player:
        return self._player2

    def get_winner(self) -> Optional[Player]:
        return self._winner

    def get_loser(self) -> Optional[Player]:
        return self._loser

    def get_board(self) -> Board:
        return self._board

    # ---------------------------------------------------------------------------
    # Game State Examination
    # ---------------------------------------------------------------------------

    def ready_to_start(self) -> bool:
        # Bug... the next line returns None
        #  return self._player1 and self._player2

        if self._player1 is None or self._player2 is None:
            return False

        return True

    def not_ready_to_start(self) -> bool:
        return not self.ready_to_start()

    def ended_with_win(self) -> bool:
        return self._winner is not None

    def ended_with_stalemate(self) -> bool:
        return self._board.is_full() and (self._winner is None)

    def is_over(self) -> bool:
        return self.ended_with_win() or self.ended_with_stalemate()

    def is_not_over(self) -> bool:
        return not self.is_over()

    def current_state(self) -> GameState:
        if self.not_ready_to_start():
            return GameState.NOT_STARTED

        if self.ended_with_win():
            return GameState.OVER_WITH_WIN

        if self.ended_with_stalemate():
            return GameState.OVER_WITH_STALEMATE

        # The only remaining possiblity is for the game to be in progress
        return GameState.IN_PROGRESS

File: Review-15-Java-Strategy-Pattern/Example-4/src/main/java/edu/odu/cs/cs330/examples/tictactoe/Game.java
package edu.odu.cs.cs330.examples.tictactoe;

import java.io.IOException;

/**
 * Orchestrates a single match of Tic-Tac-Toe.
 */
public class Game
{
    private Board board;
    private Referee ref;

    private Player player1;
    private Player player2;

    private Player winner;
    private Player loser;

    // Disable Game Default Constructor
    private Game()
    {

    }

    /**
     * Construct a Game setting player 1 and player 2.
     *
     * @param p1 the 'X' player
     * @param p2 the 'O' player
     */
    private Game(Player p1, Player p2)
    {
        this.board = new Board();
        this.ref = new Referee(this.board);

        this.player1 = p1;
        this.player2 = p2;
        this.winner = null;
        this.loser = null;
    }

    /**
     * Play one match... until either the board is full or someone gets three
     * in a row.
     */
    public void playMatch()
        throws IOException
    {
        //while(!(this.getBoard().isFull())) {
        while (this.isNotOver()) {
            if (this.playTurn(this.player1, 'X')) {
                break;
            }

            if (this.playTurn(this.player1, 'O')) {
                break;
            }
        }

        System.out.println(this.getBoard());

        if (this.endedWithWin()) {
            System.out.printf("Congratulations %s!%n", this.getWinner());
        }
    }

    /**
     * Play one half-round of Tic-Tac-Toe (i.e., one player's turn).
     *
     * @return true if the game ended during the turn (or at the start)
     *
     * @throws IOException if there is an error reading the selected move
     */
    public boolean playTurn(Player player, char symbol)
        throws IOException
    {
        if (board.isFull()) {
            return true;
        }

        System.out.println(board);
        int move = player.nextMove();

        while (this.ref.saysMoveIsNotValid(move)) {
            move = player.nextMove();
        }
        this.board.setCell(move, symbol);

        // Player won
        if(ref.checkForWin()) {
            if (symbol == 'X') {
                // Plater 1 won
                this.winner = player1;
                this.loser = player2;
            }
            else {
                // Player 2 won
                this.winner = player2;
                this.loser = player1;
            }
            return true;
        }

        return false;
    }

    public Player getPlayer1()
    {
        return this.player1;
    }

    public Player getPlayer2()
    {
        return this.player2;
    }

    public Player getWinner()
    {
        return this.winner;
    }

    public Player getLoser()
    {
        return this.loser;
    }

    public boolean endedWithWin()
    {
        return (this.winner != null);
    }

    public boolean endedWithStalemate()
    {
        return this.board.isFull()
            && (this.winner == null);
    }

    public boolean isOver()
    {
        return this.endedWithWin()
            || this.endedWithStalemate();
    }

    public boolean isNotOver()
    {
        return !this.isOver();
    }

    public Board getBoard()
    {
        return this.board;
    }

    /**
     * Orchestrator for Game construction.
     */
    public static class Builder
    {
        Player player1;
        Player player2;

        public Builder()
        {
            this.player1 = null;
            this.player2 = null;
        }

        /**
         * Add a player.
         *
         * @throws TBDException if both players have already been set.
         */
        public Builder withPlayer(final Player player)
        {
            if (this.player1 == null) {
                this.player1 = player;
            }
            else if (this.player2 == null) {
                this.player2 = player;
            }
            else {
                // throw appropriate exception
            }

            return this;
        }

        /**
         * Crate and add a human player. This handles setting up the Keyboard
         * input logic.
         *
         * @param name desired player name
         *
         * @throws TBDException if both players have already been set.
         */
        public Builder withHumanPlayer(String name)
        {
            return this.withPlayer(new Player(name, new KeyboardStrategy(name)));
        }

        /**
         * Actually create the Game.
         *
         * @throws TBDException if exactly two players have not been set.
         */
        public Game build()
        {
            if (this.player1 == null || this.player2 == null) {
                // Both players must be set
                // throw appropriate exception
            }

            return new Game(this.player1, this.player2);
        }

    }

    public static Game.Builder builder()
    {
        return new Game.Builder();
    }
}
